packages/syft/src/syft/service/code_history/code_history_service.py

```diff
@@ -0,0 +1,261 @@
+                code = user_code_service.get_by_uid(context=context, uid=uid)
+
+        elif isinstance(code, UserCode):
+            result = user_code_service.get_by_uid(context=context, uid=code.id)
+            if isinstance(result, SyftError):
+                return result
+            code = result
+
+        result_code_history_list = self.stash.get_by_service_func_name(
+            credentials=context.credentials, service_func_name=code.service_func_name
+        )
+
+        if result_code_history_list.is_err():
+            return SyftError(message=result_code_history_list.err())
+
+        code_history = None
+        code_history_list = result_code_history_list.ok()
+
+        for elem in code_history_list:
+            if elem.user_verify_key == context.credentials:
+                code_history = elem
+
+        if code_history is None:
+            code_history = CodeHistory(
+                id=UID(),
+                node_uid=context.node.id,
+                user_verify_key=context.credentials,
+                service_func_name=code.service_func_name,
+            )
+            result = self.stash.set(credentials=context.credentials, obj=code_history)
+            if result.is_err():
+                return SyftError(message=result.err())
+
+        code_history.add_code(code=code, comment=comment)
+        result = self.stash.update(credentials=context.credentials, obj=code_history)
+        if result.is_err():
+            return SyftError(message=result.err())
+
+        if final_result is None:
+            return SyftSuccess(message="Code version submit success")
+        else:
+            return final_result
+
+    @service_method(
+        path="code_history.get_all", name="get_all", roles=DATA_SCIENTIST_ROLE_LEVEL
+    )
+    def get_all(
+        self, context: AuthedServiceContext
+    ) -> Union[List[CodeHistory], SyftError]:
+        """Get a Dataset"""
+        result = self.stash.get_all(context.credentials)
+        if result.is_ok():
+            return result.ok()
+        return SyftError(message=result.err())
+
+    @service_method(
+        path="code_history.get_by_id", name="get_by_id", roles=DATA_SCIENTIST_ROLE_LEVEL
+    )
+    def get_code_by_uid(
+        self, context: AuthedServiceContext, uid: UID
+    ) -> Union[SyftSuccess, SyftError]:
+        """Get a User Code Item"""
+        result = self.stash.get_by_uid(context.credentials, uid=uid)
+        if result.is_ok():
+            code_history = result.ok()
+            return code_history
+        return SyftError(message=result.err())
+
+    @service_method(path="code_history.delete_by_id", name="delete_by_id")
+    def delete(self, context: AuthedServiceContext, uid: UID):
+        result = self.stash.delete_by_uid(context.credentials, uid)
+        if result.is_ok():
+            return result.ok()
+        else:
+            return SyftError(message=result.err())
+
+    def fetch_history_for_user(
```

Comment:
the singular naming confuses me a bit here, because it still gets all the CodeHistorie**s** but not for all users. It looks like we have an inconsistency between what we call a history in the frontend (all versions of all functions for one user), and in the backend (all versions of one function for one user)